persistence/src/commonMain/kotlin/com/quran/shared/persistence/repository/PageBookmarksRepositoryImpl.kt

```diff
@@ -0,0 +1,146 @@
+                logger.w { "Bookmark not found for deletion: page=$page" }
+                throw BookmarkNotFoundException("There's no bookmark for page #$page")
+            }
+
+            val bookmark = existingBookmarks.first()
+            if (bookmark.remote_id == null) {
+                // Local-only bookmark: delete immediately
+                logger.d { "Deleting local-only bookmark: page=$page" }
+                database.bookmarksQueries.deleteBookmarkById(bookmark.local_id)
+            } else {
+                // Synced bookmark: mark as deleted
+                logger.d { "Marking synced bookmark as deleted: page=$page" }
+                database.bookmarksQueries.setDeleted(bookmark.local_id)
+            }
+        }
+    }
+
+    override suspend fun migrateBookmarks(bookmarks: List<PageBookmark>) {
+        withContext(Dispatchers.IO) {
+            // Check if bookmarks table is empty
+            val existingBookmarks = database.bookmarksQueries.getBookmarks().executeAsList()
+            if (existingBookmarks.isNotEmpty()) {
+                throw IllegalStateException("Cannot migrate bookmarks: bookmarks table is not empty. Found ${existingBookmarks.size} bookmarks.")
+            }
+
+            // Validate that all bookmarks are from the old system (no remote IDs)
+            val bookmarksWithRemoteId = bookmarks.filter { it.remoteId != null }
+            if (bookmarksWithRemoteId.isNotEmpty()) {
+                throw IllegalArgumentException("Cannot migrate bookmarks with remote IDs. Found ${bookmarksWithRemoteId.size} bookmarks with remote IDs.")
+            }
+
+            database.bookmarksQueries.transaction {
+                bookmarks.forEach { bookmark ->
+                    database.bookmarksQueries.createLocalBookmark(bookmark.page.toLong())
+                }
+            }
+        }
+    }
+
+    override suspend fun fetchMutatedBookmarks(): List<BookmarkMutation> {
+        return withContext(Dispatchers.IO) {
+            database.bookmarksQueries.getUnsyncedBookmarks()
+                .executeAsList()
+                .map { it.toBookmarkMutation() }
+        }
+    }
+
+    override suspend fun setToSyncedState(updatesToPersist: List<BookmarkMutation>) {
+        logger.i { "Setting to synced state with ${updatesToPersist.size} updates to persist" }
+        withContext(Dispatchers.IO) {
+            // Validate all bookmarks have remote IDs
+            val invalidBookmarks = updatesToPersist.filter { it.remoteId == null }
+            if (invalidBookmarks.isNotEmpty()) {
+                logger.e { "Found ${invalidBookmarks.size} bookmarks without remote IDs" }
+                throw IllegalArgumentException("All bookmarks must have a remote ID. Found ${invalidBookmarks.size} bookmarks without remote IDs.")
+            }
+
+            database.bookmarksQueries.transaction {
+                database.bookmarksQueries.removeLocallyAddedBookmarks()
```

Comment:
this concerns me.
consider the case where:

a. I start syncing.
b. syncing takes some time (fails, retry, etc).
c. in the meanwhile, add another local bookmark.

we lose any added local bookmark.